pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java

```diff
@@ -643,6 +654,59 @@ public void initializeDispatchRateLimiterIfNeeded(Optional<Policies> policies) {
             this.dispatchRateLimiter = Optional.of(new DispatchRateLimiter(topic, name));
         }
     }
-    
+
+    @Override
+    public synchronized boolean trackDelayedDelivery(long ledgerId, long entryId, MessageMetadata msgMetadata) {
+        if (!isDelayedDeliveryEnabled) {
+            // If broker has the feature disabled, always deliver messages immediately
+            return false;
+        }
+
+        if (!delayedDeliveryTracker.isPresent()) {
+            // Initialize the tracker the first time we need to use it
+            delayedDeliveryTracker = Optional.of(new DelayedDeliveryTracker(this));
+        }
+
+        return delayedDeliveryTracker.get().addMessage(ledgerId, entryId, msgMetadata.getDeliverAtTime());
+    }
+
+    /**
+     * Returns whether we have any message that could be immediately replayed.
+     * This could be a message that was requested to be re-delivered or a delayed
+     * delivery.
+     */
+    private boolean hasMessagesToReplay() {
+        if (!messagesToRedeliver.isEmpty()) {
```

Comment:
This read of field `messagesToRedeliver` is not guarded by intrinsic lock I think it should be protected too.